packages/circuit-ui/components/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,275 @@
+      }
+      // restore scroll tp page
+      document.documentElement.style.overflowY = 'unset';
+      // trigger close animation
+      dialogElement.classList.remove(classes.show);
+      if (!hasNativeDialog) {
+        (dialogElement.nextSibling as HTMLDivElement).classList.remove(
+          classes['backdrop-visible'],
+        );
+      }
+      // trigger dialog close after animation
+      setTimeout(() => {
+        if (dialogElement.open) {
+          dialogElement.close();
+        }
+      }, animationDuration);
+    }, []);
+
+    function onPolyfillDialogKeydown(event: KeyboardEvent) {
+      if (isEscape(event) && preventClose) {
+        event.preventDefault();
+        event.stopPropagation();
+      }
+    }
+
+    function onPolyfillBackdropClick(event: MouseEvent) {
+      if (preventClose) {
+        event.preventDefault();
+      }
+    }
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+      if (!dialogElement) {
+        return undefined;
+      }
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore The package is bundled incorrectly
+      dialogPolyfill.registerDialog(dialogElement);
+      if (preventClose) {
+        dialogElement.addEventListener('keydown', onPolyfillDialogKeydown);
+      }
+      if (onClose) {
+        dialogElement.addEventListener('close', onClose);
+      }
+
+      return () => {
+        if (onClose) {
+          dialogElement.removeEventListener('close', onClose);
+        }
+        if (!hasNativeDialog && dialogElement.nextSibling) {
+          (dialogElement.nextSibling as HTMLDivElement).removeEventListener(
+            'click',
+            onPolyfillBackdropClick,
+          );
+          dialogElement.removeEventListener('keydown', onPolyfillDialogKeydown);
+        }
+      };
+    }, [onClose]);
+
+    useEffect(() => {
+      const dialogElement = dialogRef.current;
+
+      if (!dialogElement) {
+        return undefined;
+      }
+      if (open) {
+        if (document.activeElement instanceof HTMLElement) {
+          lastFocusedElementRef.current = document.activeElement;
+        }
+        if (!dialogElement.open) {
+          dialogElement.showModal();
+          if (!hasNativeDialog) {
+            // use the polyfill backdrop
+            (dialogElement.nextSibling as HTMLDivElement).classList.add(
+              classes['backdrop-visible'],
+              classes.backdrop,
+            );
+            // intercept and prevent modal closing if preventClose is true
+            (dialogElement.nextSibling as HTMLDivElement).addEventListener(
+              'click',
+              onPolyfillBackdropClick,
+            );
+          }
+
+          // trigger show animation
+          dialogElement.classList.add(classes.show);
+          // if dialog is modal, disable scroll on page
+          document.documentElement.style.overflowY = 'hidden';
```

Comment:
Unfortunately, preventing scroll isn't that simple because of browser inconsistencies (looking at you, iOS Safari 🙄). This [CSS Tricks article](https://css-tricks.com/prevent-page-scrolling-when-a-modal-is-open/) covers a holistic approach.